### ISSUE TITLE
Fine-tuning

### ISSUE THREAD

**Issue by VanitarNordic** [GeorgeSeif/Semantic-Segmentation-Suite#15]
Hello,

How can fine-tune a model on a custom dataset, with possibly a different number of classes using a pre-trained weight?

Second, which annotation style we should follow? The annotation for each datase is different, for example VOC uses white borders of `dont-care =255` which others dont.

**Comment by VanitarNordic** [comment]
Besides, I want to train and test this in Windows under an Anaconda environment. I don't know it is compatible with which Python version.

**Comment by VanitarNordic** [comment]
Fro example, I want to train on a small two class dataset (one background and another one object) by `FC-DenseNet103`, the same as you.

Typically it must be a pre-trained weights to do fine-tuning or transfer learning and then some modifications to change it for a different number of classes. I don't have equipment and mutiGPUs to train from scratch by a big dataset.

**Comment by GeorgeSeif** [comment]
Hi Vanitar,

You have a few different questions so I'll write them down here and answer them separately:

**How can you fine-tune a model on a custom dataset?**

Well as usual you will need some pre-trained weights as you mentioned in your third comment. I haven't posted those yet as I'm currently using my GPU for my own research. Once I get a chance I'll train some models and post the weights. Also keep in mind that when fine tuning your model, your new problem for which you are fine tuning should at least have some similarities to the original pre-trained one. For example, if you want to segment animals, using FC-DenseNet pretrained on CamVid would probably perform poorly.

**Which annotation style should we follow?**

You can set any custom annotation style you want in the "CamVid/class_dict.csv" file. Please see the "Usage" section of the README.

**How can I train this on Windows?**

This should work with both versions of Python, regardless of your OS. Just make sure you have all of the dependencies installed and you'll be fine (i.e Python, Numpy, TensorFlow, OpenCV).


Cheers!

**Comment by VanitarNordic** [comment]
> Well as usual you will need some pre-trained weights as you mentioned in your third comment. I haven't posted those yet as I'm currently using my GPU for my own research.

Do you have a single GPU? I don't think a single GPU would be enough for training from scratch because I have a GTX 1060 6G which only I can fine-tune some simpler models, not heavy ones. You know better, if it is possible with 6G, just let me know to try.

Another challenge with fine-tuning is that the pre-trained weight contains for example 20 classes, but the custom dataset may contain any number of classes. This must be solved either.

**Comment by GeorgeSeif** [comment]
Yes I have a 1080Ti with 11GB of RAM. I'm not sure if 6GB is enough. Why
not just try it? It'll just throw an error if it's not enough memory.

Right. So you'll have to load the weights into the original model, pop the
last layer, and then attach a new softmax layer with the desired number of
classes.

On Mar 2, 2018 2:26 PM, "Hesam Moshiri" <notifications@github.com> wrote:

Well as usual you will need some pre-trained weights as you mentioned in
your third comment. I haven't posted those yet as I'm currently using my
GPU for my own research.

Do you have a single GPU? I don't think a single GPU would be enough for
training from scratch because I have a GTX 1060 6G which only I can
fine-tune some simpler models, not heavy ones. You know better, if it is
possible with 6G, just let me know to try.

Another challenge with fine-tuning is that the pre-trained weight contains
for example 20 classes, but the custom dataset may contain any number of
classes. This must be solved either.

—
You are receiving this because you commented.

Reply to this email directly, view it on GitHub
<https://github.com/GeorgeSeif/Semantic-Segmentation-Suite/issues/15#issuecomment-370026565>,
or mute the thread

**Comment by VanitarNordic** [comment]
I have also one off-topic question.

1) Have you implemented all of these models in Python yourself alone, by reading and comparing their reference paper? 

2) if Yes, How can I be a professional like you in this? I mean what sources should I study to reach your level? These are a very complex models which you have re-implemented them.

3) Can these implementation be done by Keras? as I see you have used the TF directly.

**Comment by VanitarNordic** [comment]
>  You can set any custom annotation style you want in the "CamVid/class_dict.csv" file. 

In the VOC format, there is another layer named `dont-care`. it is the border between objects and background and should not be regarded as part of the objects, nor background. it has a white color. How should we address this? if we write as one of the classes in the CSV, it will be considered as a class, which is NOT.

![img2](https://user-images.githubusercontent.com/18719591/36933520-5d733c96-1eef-11e8-8bfe-74173402d086.png)

**Comment by GeorgeSeif** [comment]
Yup. Pretty much just reading and implementing right away.

It really just takes a lot of practice. To start off, you can always check out how other people have done it, then once you get the hang of it try one on your own!

Yes they can be done in Keras!

**Comment by GeorgeSeif** [comment]
Hhmm I'm not sure. I've never looked at the VOC data.
